File: experiments/adrianco/experiment-8/runs/language=elixir_model=claude-opus-4-7/rep1/test/book_api_test.exs
defmodule BookApi.RouterTest do
  use ExUnit.Case, async: false
  import Plug.Test
  import Plug.Conn

  alias BookApi.Router
  alias BookApi.Repo

  @opts Router.init([])

  setup do
    Repo.reset!()
    :ok
  end

  defp json_request(method, path, body \\ nil) do
    conn = conn(method, path, body && Jason.encode!(body))

    conn =
      if body do
        put_req_header(conn, "content-type", "application/json")
      else
        conn
      end

    Router.call(conn, @opts)
  end

  defp decode(conn), do: Jason.decode!(conn.resp_body)

  test "GET /health returns 200" do
    conn = json_request(:get, "/health")
    assert conn.status == 200
    assert decode(conn) == %{"status" => "ok"}
  end

  test "POST /books creates a book and GET /books/:id returns it" do
    conn =
      json_request(:post, "/books", %{
        "title" => "Dune",
        "author" => "Frank Herbert",
        "year" => 1965,
        "isbn" => "978-0441172719"
      })

    assert conn.status == 201
    created = decode(conn)
    assert created["title"] == "Dune"
    assert created["author"] == "Frank Herbert"
    assert created["year"] == 1965
    assert is_integer(created["id"])

    conn = json_request(:get, "/books/#{created["id"]}")
    assert conn.status == 200
    assert decode(conn)["title"] == "Dune"
  end

  test "POST /books returns 422 when title or author missing" do
    conn = json_request(:post, "/books", %{"author" => "Anon"})
    assert conn.status == 422
    assert decode(conn)["errors"] == ["title is required"]

    conn = json_request(:post, "/books", %{"title" => "Untitled"})
    assert conn.status == 422
    assert decode(conn)["errors"] == ["author is required"]
  end

  test "GET /books filters by author" do
    json_request(:post, "/books", %{"title" => "A", "author" => "Alice"})
    json_request(:post, "/books", %{"title" => "B", "author" => "Bob"})
    json_request(:post, "/books", %{"title" => "C", "author" => "Alice"})

    conn = json_request(:get, "/books?author=Alice")
    assert conn.status == 200
    books = decode(conn)
    assert length(books) == 2
    assert Enum.all?(books, &(&1["author"] == "Alice"))

    conn = json_request(:get, "/books")
    assert length(decode(conn)) == 3
  end

  test "PUT /books/:id updates fields and DELETE removes the book" do
    create_conn =
      json_request(:post, "/books", %{"title" => "Old Title", "author" => "Someone"})

    id = decode(create_conn)["id"]

    conn =
      json_request(:put, "/books/#{id}", %{"title" => "New Title", "year" => 2020})

    assert conn.status == 200
    updated = decode(conn)
    assert updated["title"] == "New Title"
    assert updated["author"] == "Someone"
    assert updated["year"] == 2020

    conn = json_request(:delete, "/books/#{id}")
    assert conn.status == 204

    conn = json_request(:get, "/books/#{id}")
    assert conn.status == 404
  end

  test "GET /books/:id returns 404 for missing book" do
    conn = json_request(:get, "/books/99999")
    assert conn.status == 404
    assert decode(conn)["error"] == "book not found"
  end
end

File: experiments/adrianco/experiment-6/runs/language=java_model=claude-opus-4-7_tooling=none/rep3/src/test/java/com/example/books/BookApiTest.java
package com.example.books;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class BookApiTest {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private App app;
    private Path dbFile;
    private HttpClient client;
    private String base;

    @BeforeEach
    void setUp() throws Exception {
        dbFile = Files.createTempFile("books-test-", ".db");
        Files.deleteIfExists(dbFile);
        BookRepository repo = new BookRepository("jdbc:sqlite:" + dbFile.toAbsolutePath());
        app = new App(0, repo);
        app.start();
        client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(5)).build();
        base = "http://localhost:" + app.port();
    }

    @AfterEach
    void tearDown() throws Exception {
        app.stop();
        Files.deleteIfExists(dbFile);
    }

    @Test
    void healthEndpointReturnsOk() throws Exception {
        HttpResponse<String> resp = get("/health");
        assertEquals(200, resp.statusCode());
        JsonNode body = MAPPER.readTree(resp.body());
        assertEquals("ok", body.get("status").asText());
    }

    @Test
    void createListGetUpdateDeleteBook() throws Exception {
        String payload = """
                {"title":"Dune","author":"Frank Herbert","year":1965,"isbn":"978-0441013593"}
                """;
        HttpResponse<String> created = post("/books", payload);
        assertEquals(201, created.statusCode());
        JsonNode createdBody = MAPPER.readTree(created.body());
        long id = createdBody.get("id").asLong();
        assertEquals("Dune", createdBody.get("title").asText());

        HttpResponse<String> list = get("/books");
        assertEquals(200, list.statusCode());
        JsonNode listBody = MAPPER.readTree(list.body());
        assertTrue(listBody.isArray());
        assertEquals(1, listBody.size());

        HttpResponse<String> single = get("/books/" + id);
        assertEquals(200, single.statusCode());
        assertEquals("Frank Herbert", MAPPER.readTree(single.body()).get("author").asText());

        String updated = """
                {"title":"Dune (Revised)","author":"Frank Herbert","year":1965,"isbn":"978-0441013593"}
                """;
        HttpResponse<String> upd = put("/books/" + id, updated);
        assertEquals(200, upd.statusCode());
        assertEquals("Dune (Revised)", MAPPER.readTree(upd.body()).get("title").asText());

        HttpResponse<String> del = delete("/books/" + id);
        assertEquals(204, del.statusCode());

        HttpResponse<String> gone = get("/books/" + id);
        assertEquals(404, gone.statusCode());
    }

    @Test
    void rejectsBookWithoutRequiredFields() throws Exception {
        HttpResponse<String> noTitle = post("/books", "{\"author\":\"Anon\"}");
        assertEquals(400, noTitle.statusCode());
        assertTrue(noTitle.body().contains("title"));

        HttpResponse<String> noAuthor = post("/books", "{\"title\":\"Untitled\"}");
        assertEquals(400, noAuthor.statusCode());
        assertTrue(noAuthor.body().contains("author"));
    }

    @Test
    void filterByAuthor() throws Exception {
        post("/books", "{\"title\":\"A\",\"author\":\"Alice\"}");
        post("/books", "{\"title\":\"B\",\"author\":\"Bob\"}");
        post("/books", "{\"title\":\"C\",\"author\":\"Alice\"}");

        HttpResponse<String> all = get("/books");
        assertEquals(3, MAPPER.readTree(all.body()).size());

        HttpResponse<String> alice = get("/books?author=Alice");
        JsonNode body = MAPPER.readTree(alice.body());
        assertEquals(2, body.size());
        for (JsonNode n : body) {
            assertEquals("Alice", n.get("author").asText());
        }
    }

    @Test
    void getMissingBookReturns404() throws Exception {
        HttpResponse<String> resp = get("/books/9999");
        assertEquals(404, resp.statusCode());
    }

    private HttpResponse<String> get(String path) throws Exception {
        return client.send(HttpRequest.newBuilder(URI.create(base + path)).GET().build(),
                HttpResponse.BodyHandlers.ofString());
    }

    private HttpResponse<String> post(String path, String body) throws Exception {
        return client.send(HttpRequest.newBuilder(URI.create(base + path))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body)).build(),
                HttpResponse.BodyHandlers.ofString());
    }

    private HttpResponse<String> put(String path, String body) throws Exception {
        return client.send(HttpRequest.newBuilder(URI.create(base + path))
                .header("Content-Type", "application/json")
                .PUT(HttpRequest.BodyPublishers.ofString(body)).build(),
                HttpResponse.BodyHandlers.ofString());
    }

    private HttpResponse<String> delete(String path) throws Exception {
        return client.send(HttpRequest.newBuilder(URI.create(base + path)).DELETE().build(),
                HttpResponse.BodyHandlers.ofString());
    }
}
